File: WU-Go-ui/internal/auth/token.go
package auth

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"

	"WU-Go-ui/internal/support"
)


func AcquireToken(ctx context.Context, httpClient *http.Client, tenantID, clientID, clientSecret string) (string, error) {
	u := fmt.Sprintf("https://login.microsoftonline.com/%s/oauth2/token", tenantID)

	form := url.Values{}
	form.Set("grant_type", "client_credentials")
	form.Set("client_id", clientID)
	form.Set("client_secret", clientSecret)
	form.Set("resource", "https://manage.devcenter.microsoft.com")

	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, u, strings.NewReader(form.Encode()))
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", support.NewAPIError(fmt.Sprintf("Acquire Token Failed: %d\n%s", resp.StatusCode, string(body)))
	}

	var obj map[string]any
	if err := json.Unmarshal(body, &obj); err != nil {
		return "", support.NewAPIError("Token response invalid JSON: " + err.Error())
	}

	token, _ := obj["access_token"].(string)
	if support.IsBlank(token) {
		return "", support.NewAPIError("Response missing access_token: " + string(body))
	}
	return token, nil
}

File: WU-npm/src/auth/token.ts
import { APIError } from '../support/errors.js'
import { isBlank } from '../support/strings.js'

export async function acquireToken(
  tenantId: string,
  clientId: string,
  clientSecret: string,
  opts: { signal?: AbortSignal } = {},
): Promise<string> {
  const url = `https://login.microsoftonline.com/${tenantId}/oauth2/token`
  const form = new URLSearchParams({
    grant_type: 'client_credentials',
    client_id: clientId,
    client_secret: clientSecret,
    resource: 'https://manage.devcenter.microsoft.com',
  })

  const resp = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: form.toString(),
    signal: opts.signal,
  })

  const body = await resp.text()
  if (resp.status < 200 || resp.status >= 300) {
    throw new APIError(`获取 token 失败: ${resp.status}\n${body}`)
  }

  let obj: any
  try {
    obj = JSON.parse(body)
  } catch (e) {
    throw new APIError('token 响应不是合法 JSON: ' + (e as Error).message)
  }

  const token = typeof obj.access_token === 'string' ? obj.access_token : ''
  if (isBlank(token)) throw new APIError('响应缺少 access_token: ' + body)
  return token
}
